From f055cf0ee18e5e0f7a71ef0c39095c4b1f62e5bf Mon Sep 17 00:00:00 2001
From: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Date: Thu, 6 Jun 2024 08:55:31 +0200
Subject: [PATCH 0964/1695] drivers: stm32_bsec3: fix OTP parsing to retrieve
 the permanent lock status

During initialisation, reload all OTP (even hardware shadow ones) to
register their status.
The permanent programming lock status, can be access only after reload
the OTP in the shadow.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 6f72c785daeaeb94ada9972eaa97ea0ee171f0f5 drivers: stm32_bsec3: fix OTP parsing to retrieve the permanent lock status]
Change-Id: I7e98d7b3e14eff990152e90acba68492016f4c53
Signed-off-by: Thomas Bourgoin <thomas.bourgoin@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/385692
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 core/drivers/stm32_bsec3.c | 9 ++-------
 1 file changed, 2 insertions(+), 7 deletions(-)

diff --git a/core/drivers/stm32_bsec3.c b/core/drivers/stm32_bsec3.c
index d08e2691d..d76c112e3 100644
--- a/core/drivers/stm32_bsec3.c
+++ b/core/drivers/stm32_bsec3.c
@@ -895,8 +895,6 @@ static void stm32_bsec_shadow_load(uint32_t status)
 	uint32_t srlock[OTP_ACCESS_SIZE] = { 0U };
 	uint32_t swlock[OTP_ACCESS_SIZE] = { 0U };
 	uint32_t splock[OTP_ACCESS_SIZE] = { 0U };
-	uint32_t sfsr[OTP_ACCESS_SIZE] = { 0U };
-	uint32_t otpvldr[OTP_ACCESS_SIZE] = { 0U };
 	uint32_t mask = 0U;
 	unsigned int max_id = bsec_dev.max_id;
 
@@ -920,8 +918,6 @@ static void stm32_bsec_shadow_load(uint32_t status)
 		srlock[bank] = io_read32(bsec_base() + BSEC_SRLOCK(bank));
 		swlock[bank] = io_read32(bsec_base() + BSEC_SWLOCK(bank));
 		splock[bank] = io_read32(bsec_base() + BSEC_SPLOCK(bank));
-		sfsr[bank] = io_read32(bsec_base() + BSEC_SFSR(bank));
-		otpvldr[bank] = io_read32(bsec_base() + BSEC_OTPVLDR(bank));
 	}
 
 	for (otp = 0U; otp <= max_id ; otp++) {
@@ -947,9 +943,8 @@ static void stm32_bsec_shadow_load(uint32_t status)
 			continue;
 		}
 
-		/* request shadow if not yet done or not valid */
-		if (!(sfsr[bank] & mask) || !(otpvldr[bank] & mask))
-			shadow_otp(otp);
+		/* reload shadow to read Permanent Programing Lock Flag */
+		shadow_otp(otp);
 
 		bsec_dev.shadow->value[otp] = io_read32(bsec_base() +
 					      BSEC_FVR(otp));
-- 
2.47.2

